include/lbann/layers/data_type_layer.hpp

```diff
@@ -64,44 +64,56 @@ using supported_layer_data_type = h2::meta::TL<
 #endif
   float, double>;
 
-template <typename TensorDataType>
+template <typename InputTensorDataType,
+          typename OutputTensorDataType = InputTensorDataType>
 class data_type_layer : public Layer {
 public:
   /** @name Public Types */
   ///@{
 
   /** @brief The tensor type expected in this object. */
-  using AbsDistMatrixType = El::AbstractDistMatrix<TensorDataType>;
+  using AbsDistMatrixType = El::AbstractDistMatrix<InputTensorDataType>;
+  using InputAbsDistMatrixType = El::AbstractDistMatrix<InputTensorDataType>;
+  using OutputAbsDistMatrixType = El::AbstractDistMatrix<OutputTensorDataType>;
 
   /** @brief The proxy tensor type expected in this object. */
   template <El::Device D>
-  using AbsDistMatReadProxyType = El::AbstractDistMatrixReadDeviceProxy<TensorDataType, D>;
+  using AbsDistMatReadProxyType =
+    El::AbstractDistMatrixReadDeviceProxy<InputTensorDataType, D>;
```

Comment:
Since AbsDistMatReadProxyType is just a synonym of InputAbsDistMatReadProxyType, I am thinking that we should get rid of it and only have Input and Output versions.